src/cls/rgw/cls_rgw.cc

```diff
@@ -1224,11 +1352,20 @@ int rgw_bucket_complete_op(cls_method_context_t hctx, bufferlist *in, bufferlist
 } // rgw_bucket_complete_op
 
 template <class T>
-static int write_entry(cls_method_context_t hctx, T& entry, const string& key)
+static int write_entry(cls_method_context_t hctx, T& entry, const string& key,
+                       uint32_t& reshardlog_entries, const bool is_resharding = false)
```

Comment:
i haven't tried to audit everything in cls_rgw.cc to make sure we didn't miss any writes/deletes. but assuming you covered everything, i still worry that it'll be easy to add regressions (new writes or deletes that don't update the reshard log) in the future. cc @ivancich since you help maintain this stuff

can we try to make sure that all existing calls to `cls_cxx_map_set_val()`/`cls_cxx_map_remove_key()` go through an abstraction like this `write_entry()` template? that would make it a lot easier to ensure that we keep the reshard log consistent

for this `write_entry()` function, i'd suggest taking `rgw_bucket_dir_header& header` instead of `uint32_t& reshardlog_entries, const bool is_resharding = false`

then we can add a similar helper for calls to `cls_cxx_map_remove_key()` like:
```c++
static int remove_entry(cls_method_context_t hctx, const string& key, rgw_bucket_dir_header& header)
```

Reply:
@cbodley  @ivancich  I think `write_entry()` may be enough for `cls_cxx_map_set_val()`, and I will work on `remove_entry()`. If you think `write_entry()` is not enough, I will work at it again.

Reply:
> i'd suggest taking rgw_bucket_dir_header& header instead of uint32_t& reshardlog_entries, const bool is_resharding = false

OK.